From f9b50cd48c8cca1b9c7e8b8f6611265cc62674e1 Mon Sep 17 00:00:00 2001
Message-Id: <f9b50cd48c8cca1b9c7e8b8f6611265cc62674e1.1435782705.git.mschiffer@universe-factory.net>
In-Reply-To: <704e64165f8d37c42fd13652cf8b1c0f56f37cd6.1435782705.git.mschiffer@universe-factory.net>
References: <704e64165f8d37c42fd13652cf8b1c0f56f37cd6.1435782705.git.mschiffer@universe-factory.net>
From: Sven Eckelmann <sven@narfation.org>
Date: Fri, 17 Apr 2015 19:40:28 +0200
Subject: [PATCH 2/2] batctl: Add required includes to all files

The header files could not be build indepdent from each other. This is happened
because headers didn't include the files for things they've used. This was
problematic because the success of a build depended on the knowledge about the
right order of local includes.

Also source files were not including everything they've used explicitly.
Instead they required that transitive includes are always stable. This is
problematic because some transitive includes are not obvious, depend on config
settings and may not be stable in the future.

The order for include blocks are:

 * primary headers (main.h and the *.h file of a *.c file)
 * global linux headers
 * required local headers
 * extra forward declarations for pointers in function/struct declarations

The only exceptions are linux/bitops.h and linux/if_ether.h in packet.h. This
header file is shared with userspace applications like batctl and must
therefore build together with userspace applications. The header linux/bitops.h
is not part of the uapi headers and linux/if_ether.h conflicts with the musl
implementation of netinet/if_ether.h. The maintainers rejected the use of
__KERNEL__ preprocessor checks and thus these two headers are only in main.h.
All files using packet.h first have to include main.h to work correctly.

Reported-by: Markus Pargmann <mpa@pengutronix.de>
Signed-off-by: Sven Eckelmann <sven@narfation.org>
Signed-off-by: Marek Lindner <mareklindner@neomailbox.ch>
---
 packet.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/packet.h b/packet.h
index a51ccfc..3767a86 100644
--- a/packet.h
+++ b/packet.h
@@ -20,6 +20,9 @@
 #ifndef _NET_BATMAN_ADV_PACKET_H_
 #define _NET_BATMAN_ADV_PACKET_H_
 
+#include <asm/byteorder.h>
+#include <linux/types.h>
+
 enum batadv_packettype {
 	BATADV_IV_OGM		= 0x01,
 	BATADV_ICMP		= 0x02,
-- 
2.4.5

